### PR TITLE
Implemented feature hierarchy template

### PR DESCRIPTION
**Pull request by cstrong**
New "hierarchy" report template (see https://github.com/gkushang/cucumber-html-reporter/issues/75 )
Fixed a few linter errors
Added some code comments
Made a few gratuitous changes in calculateDuration that made the code easier for me to understand, feel free to disagree

### CODE REVIEW

File: lib/reporter.js [hunk 1]
```diff
@@ -82,26 +91,53 @@ var generateReport = function (options) {
             sec > 0 ? formattedTimeStamp +=  sec + SECONDS : '';
             ms > 0 ? formattedTimeStamp +=  ms + MILLI_SECONDS : '';
 
-            return formattedTimeStamp.trim().length === 0 ? '0s' : formattedTimeStamp;
+            return formattedTimeStamp.trim().length === 0 ? '<1ms' : formattedTimeStamp;
```
Comment (gkushang) [inline lib/reporter.js]: 👍 for displaying `<1ms`

File: lib/reporter.js [hunk 2]
```diff
@@ -249,13 +292,16 @@ var generateReport = function (options) {
...
+            suite.features.summary = featuresSummary;
+
             return suite;
 
         });
 
         suite.totalTime = calculateDuration(suite.totalTime);
 
-        suite.features = featureOutput;
+        setupSubSuiteTemplates(suite);
```
Comment (gkushang) [inline lib/reporter.js]: Its breaking the existing functionality with other formatted reports when there is a hierarchy of feature files, with the templates `bootstrap` , `foundation` or `simple`. 

To try - please move the `happy.feaure` and `unhappy.features` to respective directories, e.g. features/happy/happy.feature` and pls checkout the other templates HTML reports. 

To run the tests `npm run features`

File: templates/hierarchy/piechart.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+function drawChart(chartData) {
```
Comment (gkushang) [inline templates/hierarchy/piechart.js]: To keep it DRY, Could we use the existing `piechart.js` from bootstrap directory?

File: templates/hierarchy/script.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+$(document).ready(function() {
```
Comment (gkushang) [inline templates/hierarchy/script.js]: To keep it DRY, Could we use the existing `script.js` from bootstrap directory?

**Comment by gkushang** [comment]
Nice work @cstrong ! It looks good overall. I've a suggestion in terms of view and displaying the report as a dashboard. Let me know what you think. I do have local changes that can be reviewed as a future improvements/PR.

Suggestion: Instead of wrapping all directories under `All Features`, render directly on the report (as shown in below screenshot). + the panels are aligned to pie-charts.

Reasons: One less click & can easily viewable on dashboard as overall results. Some of the users display the HTML report on a Dashboard (TV) which refreshes every few minutes. Having all features on single page would give them a nice overview.


screen shot 2017-05-30 at 3 40 52 pm

**Comment by cstrong** [comment]
hmmm you make a convincing case!  I agree the information along the top already gives you a summary, so the top "All Features" is a bit redundant.    OK I am good with it!  Let me know if you want to push your changes to this branch and merge it, or want to do it in a separate PR, or you want me to make the change.

**Comment by gkushang** [comment]
I agree. I am afraid of nesting as well, but I believe it really depends on the users. The best results can be achieved for max 2 level of nesting, and I've hardly seen the nested directories beyond 2 levels. We must have to set some exceptions in README while using this template.

However, the scenarios panels are responsive and we would need minor alignment with the `step-duration`, otherwise it looks pretty.  Below is the 2-level nesting

screen shot 2017-05-30 at 4 15 41 pm

**Comment by cstrong** [comment]
works for me.  The template only supports up to 3 levels of nesting anyway.   I think that should be enough even for very large projects.

**Pull request merged**